core/src/main/java/org/apache/calcite/plan/RelOptCluster.java

```diff
@@ -49,6 +52,7 @@
   private MetadataFactory metadataFactory;
   private final RelTraitSet emptyTraitSet;
   private RelMetadataQuery mq;
+  private Supplier<RelMetadataQuery> mqSupplier;
```

Comment:
how about `private Supplier<? extends RelMetadataQuery> mqSupplier;` ?

Reply:
Yep, `Supplier<? extends RelMetadataQuery>` sounds better.

Reply:
Nope, it is redundant.